api/graphite.go

```diff
@@ -233,16 +233,15 @@ func (s *Server) renderMetrics(ctx *middleware.Context, request models.GraphiteR
 	}
 	plan, err := expr.NewPlan(exprs, fromUnix, toUnix, mdp, stable, opts)
 	if err != nil {
+		if request.NoProxy {
+			ctx.Error(http.StatusBadRequest, "localOnly requested, but the request cant be handled locally")
+			return
+		}
```

Comment:
this doesn't make any sense to me.
firstly, if we mandate that we must have the proxy enabled any time any kind of error is triggered (from request or internal server error or any other kind), we're basically saying the graphite proxy must always be enabled.
second, we shouldn't proxy upon any kind of error, only badRequest ones.
third, there is no benefit to replacing a specific, concrete error (likely bad request with more details) with a generic one

Reply:
also note that the execution of the request happens a bit more down, when we call `s.executePlan()`

Reply:
BTW I know that it's not obvious, but there are 3 places where "bad request" may be returned:
1) `expr.ParseMany` (a bit more up): parsing of the target queries
2) `expr.NewPlan` (right here): validation of functions that the args they got matches what they support
3) `s.executePlan()`: some functions can only determine these errors at runtime. inside of their Exec function which is called during execution (e.g. `FuncAsPercent`)

Reply:
New implementation takes care of all these cases, please take a look.

> BTW I know that it's not obvious, but there are 3 places where "bad request" may be returned:
> 
>     1. `expr.ParseMany` (a bit more up): parsing of the target queries
> 
>     2. `expr.NewPlan` (right here): validation of functions that the args they got matches what they support
> 
>     3. `s.executePlan()`: some functions can only determine these errors at runtime. inside of their Exec function which is called during execution (e.g. `FuncAsPercent`)

Reply:
> this doesn't make any sense to me.
> firstly, if we mandate that we must have the proxy enabled any time any kind of error is triggered (from request or internal server error or any other kind), we're basically saying the graphite proxy must always be enabled. not desirable.

`request.NoProxy` specifies whether or not the query comes from Graphite. When it does, we should not proxy to Graphite.

> second, we shouldn't proxy upon any kind of error, only badRequest ones.(and also expr.ErrUnknownFunction as we did before)

Done in new implementation.

> third, there is no benefit to replacing a specific, concrete error (likely bad request with more details) with a generic one

That was done this was in master. Fixed in new implementation.

Reply:
> request.NoProxy specifies whether or not the query comes from Graphite. When it does, we should not proxy to Graphite.

oops my bad. makes sense!